mlflow/spacy.py

```diff
@@ -0,0 +1,189 @@
+
+    conda_env_subpath = "conda.yaml"
+    if conda_env is None:
+        conda_env = get_default_conda_env()
+    elif not isinstance(conda_env, dict):
+        with open(conda_env, "r") as f:
+            conda_env = yaml.safe_load(f)
+    with open(os.path.join(path, conda_env_subpath), "w") as f:
+        yaml.safe_dump(conda_env, stream=f, default_flow_style=False)
+
+    pyfunc.add_to_model(mlflow_model, loader_module="mlflow.spacy",
+                        data=model_data_subpath, env=conda_env_subpath)
+    mlflow_model.add_flavor(FLAVOR_NAME, spacy_version=spacy.__version__, data=model_data_subpath)
+    mlflow_model.save(os.path.join(path, "MLmodel"))
+
+
+def log_model(spacy_model, artifact_path, conda_env=None, registered_model_name=None, **kwargs):
+    """
+    Log a spacy model as an MLflow artifact for the current run.
+
+    :param spacy_model: SpaCy model to be saved.
+    :param artifact_path: Run-relative artifact path.
+    :param conda_env: Either a dictionary representation of a Conda environment or the path to a
+                      Conda environment yaml file. If provided, this decsribes the environment
+                      this model should be run in. At minimum, it should specify the dependencies
+                      contained in :func:`get_default_conda_env()`. If ``None``, the default
+                      :func:`get_default_conda_env()` environment is added to the model.
+                      The following is an *example* dictionary representation of a Conda
+                      environment::
+
+                        {
+                            'name': 'mlflow-env',
+                            'channels': ['defaults'],
+                            'dependencies': [
+                                'python=3.7.0',
+                                'pip': [
+                                    'spacy==2.2.3'
+                                ]
+                            ]
+                        }
+    :param registered_model_name: Note:: Experimental: This argument may change or be removed in a
+                                  future release without warning. If given, create a model
+                                  version under ``registered_model_name``, also creating a
+                                  registered model if one with the given name does not exist.
+    :param kwargs: kwargs to pass to ``spacy.save_model`` method.
+    """
+    Model.log(artifact_path=artifact_path, flavor=mlflow.spacy,
+              registered_model_name=registered_model_name,
+              spacy_model=spacy_model, conda_env=conda_env, **kwargs)
+
+
+def _load_model(path):
+    import spacy
+
+    path = os.path.abspath(path)
+    return spacy.load(path)
+
+
+class _SpacyModelWrapper:
+    def __init__(self, spacy_model):
+        self.spacy_model = spacy_model
+
+    def predict(self, dataframe):
+        """
+        Only works for predicting using text categorizer.
```

Comment:
I'm not too familiar with Spacy, but it seems it'd be worth documenting this limitation in the `save_model` & `log_model` docstrings? Seems like a good idea to keep the functionality here minimal in the initial PR & expand upon it later, but do you have an idea what it'd take to extend it to supporting other types of Spacy pipeline components in the future?

Reply:
It depends on how "flexible" we want to keep the predict pyfunc method. Right now I have used it only as the "Text classifier/categorizer", however when the Language/nlp object is called (see https://spacy.io/usage/spacy-101#annotations) a Doc is created. Then you can use these Doc attributes to get the relevant information.

I would say that the text categorizer predict is the one that fits more the pyfunc predict. However the loaded model can be called and generate the Doc object at anytime. I would say **the Saved and Loaded model has no limitations. Only the predict pyfunc api works for the text categorizer.**

If you wish I can add it also to save_model and log_model, however I worry that might be misleading.

Reply:
Got it, so if I understand correctly:
* Users can save & load back spaCy models without any limitations (makes sense, since `mlflow.spacy.load_model` simply loads back the original `spaCy` model object)
* However, in order to use the saved `spaCy` model with tools dependent on the python-function (pyfunc) flavor, the saved model must be a text categorizer (one thing I'd appreciate help understanding is if most common spaCy use cases only involve a text categorizer, or if this limitation would affect many spaCy use cases). 

My intuition here is that it would be strange to allow saving/logging spaCy models that can't be used with pyfunc predict.

Reply:
Yes, I think your sum up is correct. I think the text categorizer is not the only use case of spacy, as a matter of fact the different components of the pipeline are used. However, If I understood correctly the way the pyfunc predict method works here, it usually appends a column with the "prediction" at the end of the input dataframe, and I have found that among the different spaCy pipeline components the text categorizer would fit the description.

If we want, we can try to work around the different components and find a format for what spaCy is returning when the nlp object is called.
The table here: https://spacy.io/usage/spacy-101#pipelines explains fairly well how a different pipeline component is mapped to the object it creates.

My idea was to start from the text categorizer and build up from there as new use cases arise for the pyfunc.

Reply:
@arocketman got it, makes sense that it'd be nice to incrementally add support for additional pipeline components.  Could we check via `model.pipelines` that the last component of the persisted model is an instance of `spacy.pipeline.pipes.TextCategorizer`, and only generate the pyfunc flavor if that's the case? 

If the final pipeline component isn't a TextCategorizer, we can log a warning like:

```
WARNING: Generating only the spacy flavor for the provided spacy model. This means the model can be loaded back via `mlflow.spacy.load_model`, but cannot be loaded back using pyfunc APIs like `mlflow.pyfunc.load_model` or via the `mlflow models` CLI commands. MLflow will only generate the pyfunc flavor for spacy models whose final pipeline component is an instance of spacy.pipeline.pipes.TextCategorizer.
```

I think such a change would result in nicer behavior - we'll only generate a pyfunc flavor for Spacy models that can be served as pyfuncs (users won't hit issues where they attempt to run `mlflow models serve /path/to/saved/spacy/model` and the model errors out at inference time because it's not a text classifier). We can then later add support for serving other types of spacy models as pyfuncs later on.